### DISCUSSION TITLE
configuration question

### DISCUSSION THREAD

**Discussion by pas-valkov** [milvus-io/milvus#7145] [Q&A and General discussion] [answered]
Hello!
I started to explore configuration parameters of my CPU-only mishards cluster nodes and got some questions.
My goal is data reliability & cluster performance increase. As i understood, I need to pay attention firstly on these several parameters according to https://milvus.io/docs/v1.1.1/tuning.md:
1) wal.enable && storage.auto_flush_interval
I want to increase storage.auto_flush_interval. Does it mean that if mishards falls before flushing, there will be data loss? Or wal.enable: true will save my data in log files and after mishards restart finally repeats insert operations & flush data to storage?
2) cache.insert_buffer_size
I use IVF_PQ index and I am going to set cache.cache_size as ~10% of my data size, but I cant find any info about cache.insert_buffer_size. Can someone please explain how can I count it's capacity?
3) Maybe this is a silly question, but the above parameters should be the same in the configs for all milvus nodes, and these will be the parameters of the entire mishard, am I right?

**Comment by yhmo** [comment, marked as answer]
[Mishards](https://github.com/milvus-io/milvus/tree/v1.1.1/shards) is a simple distribute system with one write node and multi read nodes. There are three kinds of nodes in a mishards system:
- mishards master node: receive client requests, distribute requests to write node or read node, reduce search result
- milvus write node: response for insert/create/delete requests
- milvus read node: response for search requests

wal.enable && storage.auto_flush_interval && cache.insert_buffer_size only take effect to write node
cache.cache_size only take effect to read node

Configurations:
- auto_flush_interval defines the interval for write node flush data from insert buffer into disk. When client call insert(), the data go through the several phases:
  1. transfer from rpc
  2. write into wal log
  3. write into insert buffer
  4. flush into disk
If the master node or write node crashed between 1 & 2, the data will be lose. If the master node or write crashed after 2, the data will be safe in wal, after you restart mishards, the data will be read from wal log and flush into disk automatically.

- wal.enable can enable/disable wal. If wal.enable=false, the data from client will be send into insert buffer directly.
- cache.insert_buffer_size defines the size of buffer for insert request, default value is 1GB. Data from client will be accumulated in insert buffer, until a flush() is call or the buffer is full filled. When the buffer is full, the write node will automatically flush the data into disk.
- cache.cache_size defines the size of cache for search request, default value is 4GB. To get best search performance, you'd better set the cache_size larger than your data size. If you set cache_size = 10% of data size, the search performance will be very poor.  Mishards could deploy multiple read nodes, each read node only load a part of the data size. Assume your data size is 100GB, and there are 5 read nodes, each read node may load 10 ~ 30Gb (due to the hash ring algorithm, data is not averagely distributed), so you need to set the cache_size to be 30GB.
- In mishards example, https://github.com/milvus-io/milvus/blob/v1.1.1/shards/all_in_one_with_mysql/wr_server.yml is for write node configuration. https://github.com/milvus-io/milvus/blob/v1.1.1/shards/all_in_one_with_mysql/ro_server.yml is for read configuration. But as I mentioned, some items is only take effect to write node, some other only for read node.

**Reply by pas-valkov** [reply]
Thanks for your reply!
One thing is still not clear, the cache.cache_size parameter. So, does the information that IVF_PQ index size is ~10% of raw vector data right?
And since search uses an index, then cache.cache_size should be equal to index size - 10% of raw data. Assume I have 600Gb of raw data and, accordingly, 600*10%=60Gb index size. Hole index should be loaded into my 5 identical read-only nodes for better performance. 60/5 = 12Gb for every read-only node ideally. Taking into account not averagely distributed data, I get 18Gb cache.cache_size parameter in every read-only node config. Did I get you right?

**Reply by yhmo** [reply]
The index size of IVF_PQ is mainly decided by its parameter "m".
Use the milvus size tool to calculate index size: https://zilliz.com/sizing-tool
You can input:
```
Number of vectors = 1000000
Dimension = 512
nlist = 128
m = 32
Segment file size = 1024
```
Then you will see: 
IVF_PQ | 1954 MB | 42 MB
Change the "m", the size also changed accordingly.

That's right:
"60/5 = 12Gb for every read-only node ideally. Taking into account not averagely distributed data, I get 18Gb",

**Reply by pas-valkov** [reply]
And cache items in config are not taken into account by write node? So, it doesn't load index (that's why I devided 60 by 5, not by 5+1 in my example) and can't do search tasks?

**Reply by yhmo** [reply]
In this deployment example:
https://github.com/milvus-io/milvus/blob/v1.1.1/shards/all_in_one_with_mysql/all_in_one.yml

There is a line:
`DISCOVERY_STATIC_HOSTS: milvus_wr,milvus_ro`

The "DISCOVERY_STATIC_HOSTS" defines which node can do search tasks. In this example, both the write node and read node are specified to do search tasks. You can include write node into "DISCOVERY_STATIC_HOSTS" then it will be devided 60 by 5+1.

**Reply by pas-valkov** [reply]
Thank you!
